### PR TITLE
RFS-237 - New bridge method getBtcTransactionConfirmations

### PR DESCRIPTION
**Pull request by amendelzon**
New method `getBtcTransactionConfirmations` programmed as a hard fork from the `secondFork` (actual name TBD) fork forwards. Actual method cost still to be estimated for testnet and mainnet.

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 15 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 8 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [MerkleTreeUtils.java#L7](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java#L7): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![CRITICAL][CRITICAL] [Rskip92MerkleProofValidator.java#L20](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/validators/Rskip92MerkleProofValidator.java#L20): Unused import - co.rsk.bitcoinj.core.PartialMerkleTree. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![MAJOR][MAJOR] [Bridge.java#L561](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L561): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L563](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L563): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L565](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L565): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1153](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1153): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1155](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1155): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [BridgeUtils.java#L48](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java#L48): The Cyclomatic Complexity of this method "getStoredBlockAtHeight" is 11 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BridgeUtils.java#L59](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java#L59): Format-string method String.format(String, Object[]) called with format string "Block Height: %d must be positive or zero" wants 1 arguments but is given 2 in co.rsk.peg.BridgeUtils.getStoredBlockAtHeight(BtcBlockstoreWithCache, int) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3AVA_FORMAT_STRING_EXTRA_ARGUMENTS_PASSED)
1. ![MAJOR][MAJOR] [TestNetSecondForkConfig.java#L22](https://github.com/rsksmart/rskj/blob/3620b5ae3ba19e23703f83b3508038ba06ac9d29/rskj-core/src/main/java/org/ethereum/config/blockchain/testnet/TestNetSecondForkConfig.java#L22): This class has 6 parents which is greater than 5 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 17 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 3 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [Bridge.java#L562](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L562): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L564](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L564): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L566](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L566): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L343](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L343): The Cyclomatic Complexity of this method "registerBtcTransaction" is 17 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1159](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1159): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1161](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1161): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [BridgeUtils.java#L48](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java#L48): The Cyclomatic Complexity of this method "getStoredBlockAtHeight" is 11 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [MerkleBranch.java#L79](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/bitcoin/MerkleBranch.java#L79): Method co.rsk.peg.bitcoin.MerkleBranch.reduceFrom(Sha256Hash) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [TestNetSecondForkConfig.java#L22](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/org/ethereum/config/blockchain/testnet/TestNetSecondForkConfig.java#L22): This class has 6 parents which is greater than 5 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)
1. ![MINOR][MINOR] [Bridge.java#L575](https://github.com/rsksmart/rskj/blob/71741fcc9402d90a51d2a3a67d81d13828ce70d7/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L575): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 17 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 3 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [Bridge.java#L562](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L562): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L564](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L564): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L566](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L566): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L343](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L343): The Cyclomatic Complexity of this method "registerBtcTransaction" is 17 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1159](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1159): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1161](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1161): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [BridgeUtils.java#L48](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java#L48): The Cyclomatic Complexity of this method "getStoredBlockAtHeight" is 11 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [MerkleBranch.java#L79](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/bitcoin/MerkleBranch.java#L79): Method co.rsk.peg.bitcoin.MerkleBranch.reduceFrom(Sha256Hash) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [TestNetSecondForkConfig.java#L22](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/org/ethereum/config/blockchain/testnet/TestNetSecondForkConfig.java#L22): This class has 6 parents which is greater than 5 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)
1. ![MINOR][MINOR] [Bridge.java#L575](https://github.com/rsksmart/rskj/blob/9810e708c43516f4d4fdd80160fd9a6be5c2fb55/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L575): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * Combines two hashes (representing nodes in a merkle tree) to produce a single hash
+     * that would be the parent of these two nodes.
+     *
+     * @param left The left hand side node bytes
+     * @param right The right hand side node bytes
+     * @return
+     */
+    public Sha256Hash combineLeftRight(Sha256Hash left, Sha256Hash right) {
```
Comment (pmprete) [inline rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java]: I think it's better to just leave this function inside MerkleBranch class as its not used, and probably won't be used by other classes. If you prefer to leave it in a separated class i would sugest to set it as a static method as the helper class don't have any state, so there is no need for a Singleton

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 2]
```diff
@@ -1117,6 +1127,43 @@ public Sha256Hash getBtcBlockchainBlockHashAtDepth(int depth) throws BlockStoreE
...
+            if (!blockAtHeight.getHeader().getHash().equals(btcBlockHash)) {
+                return BTC_TRANSACTION_CONFIRMATION_BLOCK_NOT_IN_BEST_CHAIN_ERROR_CODE;
+            }
+
+            if (!merkleBranch.proves(btcTxHash, block.getHeader())) {
+                return BTC_TRANSACTION_CONFIRMATION_INVALID_MERKLE_BRANCH_ERROR_CODE;
+            }
+
+            return this.btcBlockChain.getBestChainHeight() - block.getHeight() + 1;
+        } catch (InvalidBlockHeightException | InvalidBlockHashException e) {
```
Comment (pmprete) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: InvalidBlockHeightException is used twice

File: rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java (hunk 1 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java]: It is used in two different classes: `Rskip92MerkleProofValidator` and `MerkleBranch`.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 2 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: I believe one of them is `InvalidBlockHashException`

File: rskj-core/src/main/java/co/rsk/peg/Bridge.java [hunk 3]
```diff
@@ -549,6 +554,28 @@ public Integer getBtcBlockchainInitialBlockHeight(Object[] args)
         return blockHash.getBytes();
     }
 
+    public int getBtcTransactionConfirmations(Object[] args)
+    {
+        logger.trace("getBtcTransactionConfirmations");
+        try {
+            Sha256Hash btcTxHash = Sha256Hash.wrap((byte[]) args[0]);
```
[the remark this answers is not in the document]
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/Bridge.java]: I think it really depends on the method. We have different approaches. For instance, `registerBtcTransaction`, `addSignature`, `addFederatorPublicKey` and `commitFederation` all use `byte[]`; whereas `isBtcTxHashAlreadyProcessed` and `getBtcTxHashProcessedHeight` use `string`. I believe the "correct" way is to use the data type closest to what that we are implementing. In this case, that would be `byte[]`. Hence the decision. 

On a side note, I actually believe the cases of `isBtcTxHashAlreadyProcessed` and `getBtcTxHashProcessedHeight` were my mistakes at the time of writing when I didn't know we could just use `byte[]` for those cases.

File: rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java (hunk 1 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java]: Will change to static.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 2 above)
Comment (pmprete) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: true, sorry i miss read them

**Comment by amendelzon** [comment]
Need to address some issues within the `BridgeUtils.getStoredBlockAtHeight` method. Closing for now.

**Comment by amendelzon** [comment]
Fixed, reopening.

**Comment by rskops** [comment]
SonarQube analysis reported 11 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [MerkleTreeUtils.java#L7](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/utils/MerkleTreeUtils.java#L7): Add a private constructor to hide the implicit public one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [Bridge.java#L562](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L562): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L564](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L564): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L566](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L566): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L344](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L344): The Cyclomatic Complexity of this method "registerBtcTransaction" is 17 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TestNetSecondForkConfig.java#L22](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/org/ethereum/config/blockchain/testnet/TestNetSecondForkConfig.java#L22): This class has 6 parents which is greater than 5 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)
1. ![MINOR][MINOR] [Bridge.java#L575](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L575): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MINOR][MINOR] [MerkleBranch.java#L43](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/bitcoin/MerkleBranch.java#L43): Return a copy of "hashes". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2384)
1. ![INFO][INFO] [Bridge.java#L575](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L575): Unconstrained method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) converts checked exception to unchecked [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CONSTRAINTS)
1. ![INFO][INFO] [BridgeMethods.java#L156](https://github.com/rsksmart/rskj/blob/ddf806f55acad856711a7ab914759d77c26efe40/rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java#L156): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 10 issues

1. ![MAJOR][MAJOR] [Bridge.java#L562](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L562): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L564](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L564): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L566](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L566): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L344](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L344): The Cyclomatic Complexity of this method "registerBtcTransaction" is 17 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [TestNetSecondForkConfig.java#L22](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/org/ethereum/config/blockchain/testnet/TestNetSecondForkConfig.java#L22): This class has 6 parents which is greater than 5 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)
1. ![MINOR][MINOR] [Bridge.java#L575](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L575): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MINOR][MINOR] [MerkleBranch.java#L43](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/bitcoin/MerkleBranch.java#L43): Return a copy of "hashes". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2384)
1. ![INFO][INFO] [Bridge.java#L575](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L575): Unconstrained method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) converts checked exception to unchecked [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CONSTRAINTS)
1. ![INFO][INFO] [BridgeMethods.java#L156](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java#L156): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)
1. ![INFO][INFO] [AbstractConfig.java#L175](https://github.com/rsksmart/rskj/blob/e93100b0267a34f8f4d01d71be2918aa98cf9b8a/rskj-core/src/main/java/org/ethereum/config/blockchain/AbstractConfig.java#L175): Complete the task associated to this TODO comment. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**